README.md
# Animated Bottom Navigation

[![pub package](https://img.shields.io/pub/v/bottom_animation?color=green&style=flat-square)](https://pub.dev/packages/bottom_animation)

##### Flutter `package` for `add bottomNavigation` to your application

![pic](https://raw.githubusercontent.com/mahmoud-eslami/bottom_animation/master/files/BOTTOM%20ANIMATION.png)

## Contribution

### - Your Pull Requests are welcome 🥳🙏.

### Show case :

|                     Flat Navigation Bar Custom widget                      |                      With border radius Custom Widget                     |                           Curved with custom widget                           |
| :----------------------------------------------------------: | :----------------------------------------------------------: | :----------------------------------------------------------: |
| ![green](https://raw.githubusercontent.com/mahmoud-eslami/bottom_animation/master/files/ezgif.com-gif-maker3.gif) | ![pink](https://raw.githubusercontent.com/mahmoud-eslami/bottom_animation/master/files/ezgif.com-gif-maker%20(1).gif) | ![orange](https://raw.githubusercontent.com/mahmoud-eslami/bottom_animation/master/files/ezgif.com-gif-maker%20(2).gif) |
|                     Flat Navigation Bar                      |                      With border radius                      |                           Dark One                           |
| :----------------------------------------------------------: | :----------------------------------------------------------: | :----------------------------------------------------------: |
| ![purpule](https://github.com/mahmoud-eslami/bottom_animation/blob/master/files/purple-showCase.gif) | ![blue_grayy](https://raw.githubusercontent.com/mahmoud-eslami/bottom_animation/master/files/ezgif.com-gif-maker.gif) | ![dark](https://github.com/mahmoud-eslami/bottom_animation/blob/master/files/black-showCase.gif) |



### How to use :

#### 1. Depend on it

Add this to your package's pubspec.yaml file:

```yaml
dependencies:
  bottom_animation: ^<last_version>
```

#### 2. Install it

You can install packages from the command line:

with Flutter:

```shell
$ flutter pub get
```

Alternatively, your editor might support `flutter pub get`. Check the docs for your editor to learn more.

#### 3. Import it

Now in your Dart code, you can use:

```dart
import 'package:bottom_animation/bottom_animation.dart';
```



### Widget parameters :

|          name          |        type         |                description                 |                        default_value                         |
| :--------------------: | :-----------------: | :----------------------------------------: | :----------------------------------------------------------: |
|     selectedIndex      |         int         |           index of selected item           |                              -                               |
| hoverAlignmentDuration |         int         |      controll movement speed of hover      |                      700(milliesecond)                       |
|         items          | List<BottomNavItem> |       list of bottom navigition item       |                              -                               |
|    backgroundColor     |        Color        |   background color of bottom navigation    |                              -                               |
|    activeIconColor     |        Color        |           color of selected item           |                              -                               |
|   deactiveIconColor    |        Color        |          color of unselected item          |                              -                               |
|        iconSize        |       double        |           size of BottomNavIcon            |                              30                              |
|       textStyle        |      TextStyle      | text style for title of each BottomNavItem | TextStyle(color: Colors.white, fontSize: 18.0, fontWeight: FontWeight.w300) |
|      onItemSelect      |  ValueChanged<int>  |                     -                      |                              -                               |
|       barHeight        |       double        |          bottom navigation height          |                              80                              |
|       barRadius        |       double        |      bottom navigation border radius       |                              0                               |
|     itemHoverColor     |        Color        |       background color of each item        |                              -                               |
| itemHoverColorOpacity  |       double        |                     -                      |                              13                              |
| itemHoverBorderRadius  |       double        |                     -                      |                              15                              |
|     itemHoverWidth     |       double        |                     -                      |                             150                              |
|    itemHoverHeight     |       double        |                     -                      |                              55                              |

### BottomNavItem :

|   name   |   type   |
| :------: | :------: |
|  title   |  String  |
| iconData | IconData |




### example :

```dart
class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  var items = <BottomNavItem>[
    BottomNavItem(title: 'Home', iconData: CupertinoIcons.heart),
    BottomNavItem(title: 'Profile', iconData: CupertinoIcons.person),
    BottomNavItem(title: 'Setting', iconData: CupertinoIcons.search),
    BottomNavItem(title: 'tools', iconData: CupertinoIcons.bluetooth),
  ];

  var cIndex;
  @override
  void initState() {
    cIndex = 0;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'example',
      theme: ThemeData(
        primarySwatch: Colors.indigo,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: Scaffold(
        backgroundColor: Colors.white,
        bottomNavigationBar: BottomAnimation(
          selectedIndex: cIndex,
          items: items,
          backgroundColor: Colors.blueGrey,
          onItemSelect: (value) {
            setState(() {
              cIndex = value;
            });
          },
          itemHoverColor: Colors.white,
          itemHoverColorOpacity: .9,
          activeIconColor: Colors.blueGrey,
          deactiveIconColor: Colors.white.withOpacity(.9),
          barRadius: 40,
          textStyle: TextStyle(
            color: Colors.blueGrey,
            fontWeight: FontWeight.bold,
          ),
          itemHoverWidth: 130,
          itemHoverBorderRadius: 40,
        ),
      ),
    );
  }
}
```





CHANGELOG.md
## [1.0.0] - 2020-07-03.

 - Initial Release .

## [1.0.1] - 2020-07-03.

 - Readme Updated .

## [1.0.2] - 2020-07-10.

 - Description updated.

 ## [1.1.0] - 2020-07-15

 - Add hoverDuration field to widget constractor .
    - you can control movement speed .

 ## [1.2.0] - 2020-07-15

 - Now you can use both iconData and widget for items in bottomNavigation.

 ## [2.0.0] - 2021-06-20

 - Null safety added

test/bottomanimation_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:bottom_animation/bottom_animation.dart';

void main() {
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';
import 'package:bottom_animation/bottom_animation.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  var items = <BottomNavItem>[
    BottomNavItem(
        title: 'logo',
        widget: Image.asset(
          'assets/images/logo.png',
          width: 30,
        )),
    BottomNavItem(
        title: 'profile',
        widget: Image.asset('assets/images/profilen.png', width: 30)),
    BottomNavItem(
        title: 'star',
        widget: Image.asset('assets/images/star.png', width: 30)),
  ];

  var cIndex;
  @override
  void initState() {
    cIndex = 0;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'example',
      theme: ThemeData(
        primarySwatch: Colors.indigo,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: Scaffold(
        backgroundColor: Colors.white,
        bottomNavigationBar: BottomAnimation(
          selectedIndex: cIndex,
          items: items,
          backgroundColor: Color(0xffea8c55),
          onItemSelect: (value) {
            setState(() {
              cIndex = value;
            });
          },
          itemHoverColor: Color(0xfff5dd90),
          itemHoverColorOpacity: .5,
          activeIconColor: Colors.black,
          deActiveIconColor: Colors.black38,
          barRadius: 30,
          textStyle: TextStyle(
            color: Colors.black,
            fontWeight: FontWeight.bold,
            fontSize: 18,
          ),
          itemHoverWidth: 135,
          itemHoverBorderRadius: 30,
        ),
      ),
    );
  }
}


lib/source/navigation_bar.dart
import 'dart:ui';

import 'package:bottom_animation/bottom_animation.dart';
import 'package:flutter/material.dart';

/// Bottom Navigation Widget.
class BottomAnimation extends StatefulWidget {
  // widget parameters
  final int selectedIndex;
  final List<BottomNavItem> items;
  final Color backgroundColor;
  final Color activeIconColor;
  final Color deActiveIconColor;
  final double iconSize;
  final TextStyle? textStyle;
  final ValueChanged<int> onItemSelect;
  final double barHeight;
  final double barRadius;
  final Color itemHoverColor;
  final double itemHoverColorOpacity;
  final double itemHoverBorderRadius;
  final double itemHoverWidth;
  final double itemHoverHeight;
  final int hoverAlignmentDuration;

  const BottomAnimation({
    Key? key,
    required this.selectedIndex,
    required this.items,
    required this.activeIconColor,
    required this.deActiveIconColor,
    required this.backgroundColor,
    required this.onItemSelect,
    required this.itemHoverColor,
    this.hoverAlignmentDuration = 700,
    this.iconSize = 30,
    this.textStyle,
    this.barHeight = 80,
    this.barRadius = 0,
    this.itemHoverBorderRadius = 15,
    this.itemHoverColorOpacity = .13,
    this.itemHoverHeight = 55,
    this.itemHoverWidth = 150,
  }) : super(key: key);

  @override
  _BottomAnimationState createState() => _BottomAnimationState();
}

class _BottomAnimationState extends State<BottomAnimation> {
  var textStyle;
  late List<BottomNavItem> listItems;

  double calculateContainerPosition(int index) {
    //lerp parameter
    var isLtr = Directionality.of(context) == TextDirection.ltr;
    var listSize = widget.items.length;
    var a = 0.0;
    var b = 0.0;
    // set suitable count for calculate lerp
    if (listSize == 0) {
      a = 0.0;
      b = 0.0;
    } else if (listSize == 2) {
      a = .6;
      b = .6;
    } else if (listSize == 3) {
      a = .8;
      b = .8;
    } else if (listSize == 4) {
      a = .9;
      b = .9;
    } else {
      a = 1;
      b = 1;
    }
    // return calculated lerp
    if (isLtr)
      return lerpDouble(-a, b, index / (listSize - 1))!;
    else
      return lerpDouble(b, -a, index / (listSize - 1))!;
  }

  @override
  void initState() {
    listItems = widget.items;

    textStyle = widget.textStyle ??
        TextStyle(
          color: Colors.white,
          fontSize: 18.0,
          fontWeight: FontWeight.w300,
        );
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      decoration: BoxDecoration(
        color: widget.backgroundColor,
        borderRadius: BorderRadius.vertical(
          top: Radius.circular(
            widget.barRadius,
          ),
        ),
      ),
      width: double.infinity,
      height: widget.barHeight,
      child: Stack(
        children: <Widget>[
          AnimatedAlign(
            curve: Curves.ease,
            duration: Duration(milliseconds: widget.hoverAlignmentDuration),
            alignment: Alignment(calculateContainerPosition(widget.selectedIndex), 0),
            child: Container(
              width: widget.itemHoverWidth,
              height: widget.itemHoverHeight,
              decoration: BoxDecoration(
                color: widget.itemHoverColor.withOpacity(widget.itemHoverColorOpacity),
                borderRadius: BorderRadius.all(
                  Radius.circular(widget.itemHoverBorderRadius),
                ),
              ),
            ),
          ),
          Padding(
            padding: EdgeInsets.symmetric(horizontal: 20),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: listItems.map((item) {
                var index = listItems.indexOf(item);
                return GestureDetector(
                  onTap: () => widget.onItemSelect(index),
                  child: BarItem(
                    selected: widget.selectedIndex == index,
                    activeColor: widget.activeIconColor,
                    deActiveColor: widget.deActiveIconColor,
                    widget: item.widget,
                    iconData: item.iconData,
                    title: item.title,
                    iconSize: widget.iconSize,
                    textStyle: textStyle,
                  ),
                );
              }).toList(),
            ),
          ),
        ],
      ),
    );
  }
}

/// Each item in Bottom Navigation
class BarItem extends StatefulWidget {
  final Widget? widget;
  final IconData? iconData;
  final String title;
  final bool selected;
  final Color activeColor;
  final Color deActiveColor;
  final double iconSize;
  final TextStyle textStyle;

  const BarItem({
    Key? key,
    this.widget,
    this.iconData,
    required this.title,
    required this.selected,
    required this.activeColor,
    required this.deActiveColor,
    required this.iconSize,
    required this.textStyle,
  }) : super(key: key);

  @override
  _BarItemState createState() => _BarItemState();
}

class _BarItemState extends State<BarItem> with TickerProviderStateMixin {
  @override
  void initState() {
    super.initState();
  }

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedContainer(
      curve: Curves.ease,
      width: widget.selected ? 100 : 50,
      duration: Duration(seconds: 1),
      child: SingleChildScrollView(
        scrollDirection: Axis.horizontal,
        child: Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            widget.widget ??
                Icon(
                  widget.iconData,
                  color: widget.selected ? widget.activeColor : widget.deActiveColor,
                  size: widget.iconSize,
                ),
            SizedBox(
              width: 10,
            ),
            widget.selected
                ? Text(
                    widget.title,
                    style: widget.textStyle,
                  )
                : Container()
          ],
        ),
      ),
    );
  }
}
